jax_distribution/systems/ff_ippo_rware.py

```diff
@@ -0,0 +1,407 @@
+                ) -> Tuple:
+                    """Calculate the loss."""
+                    # RERUN NETWORK
+                    actor_policy, value = apply_fn(params, traj_batch.obs)
+                    log_prob = actor_policy.log_prob(traj_batch.action)
+
+                    # CALCULATE VALUE LOSS
+                    value_pred_clipped = traj_batch.value + (
+                        value - traj_batch.value
+                    ).clip(-config["CLIP_EPS"], config["CLIP_EPS"])
+                    value_losses = jnp.square(value - targets)
+                    value_losses_clipped = jnp.square(value_pred_clipped - targets)
+                    value_loss = (
+                        0.5 * jnp.maximum(value_losses, value_losses_clipped).mean()
+                    )
+
+                    # CALCULATE ACTOR LOSS
+                    ratio = jnp.exp(log_prob - traj_batch.log_prob)
+                    gae = (gae - gae.mean()) / (gae.std() + 1e-8)
+                    loss_actor1 = ratio * gae
+                    loss_actor2 = (
+                        jnp.clip(
+                            ratio,
+                            1.0 - config["CLIP_EPS"],
+                            1.0 + config["CLIP_EPS"],
+                        )
+                        * gae
+                    )
+                    loss_actor = -jnp.minimum(loss_actor1, loss_actor2)
+                    loss_actor = loss_actor.mean()
+                    entropy = actor_policy.entropy().mean()
+
+                    total_loss = (
+                        loss_actor
+                        + config["VF_COEF"] * value_loss
+                        - config["ENT_COEF"] * entropy
+                    )
+                    return total_loss, (value_loss, loss_actor, entropy)
+
+                grad_fn = jax.value_and_grad(_loss_fn, has_aux=True)
+                total_loss, grads = grad_fn(
+                    params, opt_state, traj_batch, advantages, targets
+                )
+
+                grads, total_loss = jax.lax.pmean(
+                    (grads, total_loss), axis_name="batch"
+                )
+                grads, total_loss = jax.lax.pmean(
+                    (grads, total_loss), axis_name="device"
+                )
+
+                updates, new_opt_state = update_fn(grads, opt_state)
+                new_params = optax.apply_updates(params, updates)
+
+                return (new_params, new_opt_state), total_loss
+
+            params, opt_state, traj_batch, advantages, targets, rng = update_state
+            rng, _rng = jax.random.split(rng)
+
+            # SHUFFLE MINIBATCHES
+            batch_size = config["ROLLOUT_LENGTH"] * config["NUM_ENVS"]
+            permutation = jax.random.permutation(_rng, batch_size)
+            batch = (traj_batch, advantages, targets)
+            batch = jax.tree_util.tree_map(
+                lambda x: x.reshape((batch_size,) + x.shape[2:]), batch
+            )
```

Comment:
This is done quite a lot in general, I know [haiku has](https://github.com/deepmind/dm-haiku/blob/main/haiku/_src/basic.py#L207) `merge_leading_dims` I couldn't find it for flax. If it definitely doesn't have this then we should add our own implementation in a utils file so that we can do:

```python
batch = jax.tree_util.tree_map(lambda x: merge_leading_dims(x, 2), batch)
```